official/projects/lra/lra_dual_encoder.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Trainer network for dual encoder style models."""
+# pylint: disable=g-classes-have-attributes
+import collections
+import tensorflow as tf
+
+from official.nlp.modeling import layers
+
+
+@tf.keras.utils.register_keras_serializable(package='Text')
+class LRADualEncoder(tf.keras.Model):
+  """A dual encoder model based on a transformer-based encoder.
+
+  This is an implementation of the dual encoder network structure based on the
+  transfomer stack, as described in ["Language-agnostic BERT Sentence
+  Embedding"](https://arxiv.org/abs/2007.01852)
+
+  The DualEncoder allows a user to pass in a transformer stack, and build a dual
+  encoder model based on the transformer stack.
+
+  Args:
+    network: A transformer network which should output an encoding output.
+    max_seq_length: The maximum allowed sequence length for transformer.
+    normalize: If set to True, normalize the encoding produced by transfomer.
+    logit_scale: The scaling factor of dot products when doing training.
+    logit_margin: The margin between positive and negative when doing training.
+    output: The output style for this network. Can be either `logits` or
+      `predictions`. If set to `predictions`, it will output the embedding
+      producted by transformer network.
+  """
+
+  def __init__(self,
+               network,
+               num_classes,
+               max_seq_length,
+               dropout_rate=0.1,
+               initializer='glorot_uniform',
+               use_encoder_pooler=True,
+               head_name='dual_encode',
+               **kwargs) -> None:
+
+    left_word_ids = tf.keras.layers.Input(shape=(max_seq_length,),
```

Comment:
Can we make it a subclass Keras layer: e.g.. using `call()` to have these logics.

Reply:
This part was following `official/nlp/modeling/models/dual_encoder.py` do you have an example for the keras layer style? Be happy to adapt to that

Reply:
Hi Evan, you can take a look at: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/networks/bert_encoder.py#L223
The BertEncoderV2 is modified to be a keras layer and using call() override.

Reply:
Keras layer and call() rewrite finished